mypy/semanal.py

```diff
@@ -3936,11 +3936,26 @@ class C:
         #       caught.
         assert self.statement  # we are at class scope
         return (node is None
-                or node.line < self.statement.line
+                or self.is_textually_before_statement(node)
                 or not self.is_defined_in_current_module(node.fullname)
                 or isinstance(node, TypeInfo)
                 or (isinstance(node, PlaceholderNode) and node.becomes_typeinfo))
 
+    def is_textually_before_statement(self, node: SymbolNode) -> bool:
+        """Check if a node is defined textually before the current statement
+
+        Note that decorated functions' line number are the same as
+        the top decorator.
+        """
+        assert self.statement
+        line_diff = self.statement.line - node.line
+        if isinstance(node, OverloadedFuncDef) and isinstance(self.statement, FuncDef):
+            return node.name != self.statement.name and line_diff > 0
```

Comment:
Comparing names looks like a hack as it can confuse two totally unrelated definitions. You can instead directly check if function definition is one of the overload items (the latter are stored as `.items` and `.impl`). This can be written as a separate helper method.

Also add a comment that this is a corner case where mypy technically deviates from runtime name resolution, but it is fine because we want an overloaded function to be treated as a single unit.